Avoid open-ended dependencies, use lock file

avdv opened the pull request:
### Description

Running `rake install` always resulted in these warnings:
```
WARNING:  open-ended dependency on clocale (>= 0) is not recommended
  if clocale is semantically versioned, use:
    add_runtime_dependency 'clocale', '~> 0'
WARNING:  open-ended dependency on filesize (>= 0) is not recommended
  if filesize is semantically versioned, use:
    add_runtime_dependency 'filesize', '~> 0'
WARNING:  open-ended dependency on manpages (>= 0) is not recommended
  if manpages is semantically versioned, use:
    add_runtime_dependency 'manpages', '~> 0'
WARNING:  open-ended dependency on rainbow (>= 0) is not recommended
  if rainbow is semantically versioned, use:
    add_runtime_dependency 'rainbow', '~> 0'
WARNING:  open-ended dependency on diffy (>= 0, development) is not recommended
  if diffy is semantically versioned, use:
    add_development_dependency 'diffy', '~> 0'
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  open-ended dependency on ronn (>= 0, development) is not recommended
  if ronn is semantically versioned, use:
    add_development_dependency 'ronn', '~> 0'
WARNING:  open-ended dependency on rspec (>= 0, development) is not recommended
  if rspec is semantically versioned, use:
    add_development_dependency 'rspec', '~> 0'
WARNING:  open-ended dependency on rspec-its (>= 0, development) is not recommended
  if rspec-its is semantically versioned, use:
    add_development_dependency 'rspec-its', '~> 0'
WARNING:  open-ended dependency on rubocop (>= 0, development) is not recommended
  if rubocop is semantically versioned, use:
    add_development_dependency 'rubocop', '~> 0'
WARNING:  open-ended dependency on rubocop-rspec (>= 0, development) is not recommended
  if rubocop-rspec is semantically versioned, use:
    add_development_dependency 'rubocop-rspec', '~> 0'
WARNING:  open-ended dependency on rubygems-tasks (>= 0, development) is not recommended
  if rubygems-tasks is semantically versioned, use:
    add_development_dependency 'rubygems-tasks', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
```

- remove Gemfile.lock from gitignore and add it to the repository
- use pessimistic version constraints for all dependencies in the gemspec

- Relevant Issues : (none)
- Relevant PRs : (none)
- Type of change :
  - [ ] New feature
  - [ ] Bug fix for existing feature
  - [x] Code quality improvement
  - [ ] Addition or Improvement of tests
  - [ ] Addition or Improvement of documentation

athityakumar commented:
Won't adding the pessimistic versions make the warnings disappear? Is it necessary to check the Gemfile.lock into git? It's generally not included in the repository for Ruby gems.

avdv commented:
> Won't adding the pessimistic versions make the warnings disappear? Is it necessary to check the Gemfile.lock into git?

The warnings are one thing, but it would be necessary to avoid a dependency update causing the build to break (for example when rubocop gets a new cop).

http://bundler.io/v1.7/rationale.html#checking-your-code-into-version-control

> This is important: **the Gemfile.lock makes your application a single package of both your own code and the third-party code it ran the last time you know for sure that everything worked.** Specifying exact versions of the third-party code you depend on in your Gemfile would not provide the same guarantee, because gems usually declare a range of versions for their dependencies. 

Another good read: http://yehudakatz.com/2011/05/30/gem-versioning-and-bundler-doing-it-right/

avdv commented:
After reading [Clarifying the Roles of the .gemspec and Gemfile](https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/) I now understand the difference between a "gem" and an "app".

My misunderstanding was that gems are "libraries" and apps are "applications" (executables) instead. But the difference is just about how you distribute them, no matter if you have an executable in your gem or not -- you're distributing a gem, with its (inherent) shortcoming that it does not provide the same fixed version guarantees as bundler.

So, I removed the lock file, updated the version and rebased against master.

avdv commented:
@athityakumar BTW, it seems you do not have much time for this project lately. Hope you're doing well!

avdv commented:
Hey @athityakumar any progress on this? What do you think, can you merge this PR?
